### PR TITLE
patch for servers not supporting certain assumed postgresql capabilities

### PR DESCRIPTION
**Pull request by LLuke**
This issue is described in [#10656](https://github.com/sqlalchemy/sqlalchemy/issues/10656) with test code there. It is a minimal functionality fix. A caveat is that this fix doesn't deal with ramifications of handling failure of _set_backslash_escapes() which was not in the original sqlalchemy code base either.

Due to otherwise needs to setup an external database, unit tests were not added besideds the test code in the issue.  It is not expected to break any current test cases. It is recommended that sqlacademy to add mocks for all assumptions and handle them as gracefully as possible.

sqlalchemy is widely used and we hope that it is even more widely used. :-)

### Checklist

This pull request is:

- [ ] A documentation / typographical / small typing error fix
	- Good to go, no issue or tests are needed
- [X ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

File: lib/sqlalchemy/engine/default.py [hunk 1]
```diff
@@ -514,20 +514,20 @@ def initialize(self, connection):
             self.server_version_info = self._get_server_version_info(
                 connection
             )
-        except NotImplementedError:
+        except :
```
Comment (zzzeek) [inline lib/sqlalchemy/engine/default.py:517]: we can't suppress all errors here, those would need to be propagated

File: lib/sqlalchemy/dialects/postgresql/base.py [hunk 2]
```diff
@@ -3049,15 +3049,18 @@ def initialize(self, connection):
...
+        except:
+            pass
 
-        self._supports_drop_index_concurrently = self.server_version_info >= (
+        self._supports_drop_index_concurrently = self.server_version_info and self.server_version_info >= (
             9,
             2,
         )
-        self.supports_identity_columns = self.server_version_info >= (10,)
+        self.supports_identity_columns = self.server_version_info and self.server_version_info >= (10,)
```
Comment (zzzeek) [inline lib/sqlalchemy/dialects/postgresql/base.py:3063]: server_version_info must be provided here, it's invalid state for this to not be present

File: lib/sqlalchemy/dialects/postgresql/base.py [hunk 3]
```diff
@@ -3049,15 +3049,18 @@ def initialize(self, connection):
         super().initialize(connection)
 
         # https://www.postgresql.org/docs/9.3/static/release-9-2.html#AEN116689
-        self.supports_smallserial = self.server_version_info >= (9, 2)
+        self.supports_smallserial = self.server_version_info and self.server_version_info >= (9, 2)
 
-        self._set_backslash_escapes(connection)
+        try: 
+            self._set_backslash_escapes(connection)
+        except:
```
Comment (zzzeek) [inline lib/sqlalchemy/dialects/postgresql/base.py:3056]: this would cause silent errors

**Review by zzzeek — COMMENTED** [review]
overall per discussion in #10656 this is not something we do on our end

File: lib/sqlalchemy/dialects/postgresql/base.py (hunk 3 above)
Comment (LLuke) [reply lib/sqlalchemy/dialects/postgresql/base.py:3056]: Agree., also my concern when I hacked it to hear from you. 

A database should better comply but I'm curious whether sqlalchemy has a fallback solution when a database for whatever reason doesn't support this. Is there a legacy solution in history or do we know for sure where needs to be retooled for people to support that dialect without the trials and errors?

Comment (CaselIT) [reply lib/sqlalchemy/dialects/postgresql/base.py:3056]: the "this" here is custom code that's written to support postgresql.

If a database does need to do something different, the dialect can decide to handle it in a different way.

**Pull request closed without merging**